### PR TITLE
[sigh] Fix resign for widget (or other plugins)

### PR DESCRIPTION
**Pull request by MiB3**
🔑

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

When resigning an ios app which has a widget without an extra app entitlement and using `--use_app_entitlements`, the resigning fails with the following error. I don't know if there are actually ios apps where the widget has its own app entitlements file.
```
Executable=/.../_floatsignTemp/Payload/myApp.app/PlugIns/Widget.appex/Widget
...
_floatsignTemp/patchedEntitlements: unrecognized blob type (accepting blindly)
_floatsignTemp/patchedEntitlements: invalid length in entitlement blob
Encountered an error, aborting!
```
This error occurs inside `sigh/lib/assets/resign.sh` because it expects an non empty app entitlements file. The main app has one. But the Widget (which is inside the plugins folder within the ipa) does not.

### Description

More specifically the error occurs on on line 829 (https://github.com/fastlane/fastlane/blob/master/sigh/lib/assets/resign.sh#L829), because of accumulated error / failed attempts to parse the app entitlements.
I moved the parts which depend on the app entitlements into an _if_, so they are not executed if no app entitlements file is currently present. The patched entitlements file however is still generated.

I tested the changes on my app, which failed to be resigned. Now the resigning works and the app can be successfully deployed (to testers via ad-hoc).

### Testing Steps

I am sorry I can not provide an example app. I am also not sure how to add a test to fastlane which tests the resigning of such a condition. This would be really good to avoid this feature to be broken again in the future. It is also not enough to only check if the resigning ended successfully. One also needs to deploy the app to an actual device to ensure the structure and codesigning is valid.

### CODE REVIEW

**Comment by janpio** [comment]
Thanks, these changes look safe enough although I have _no_ idea what exactly is going on and how we could properly test this :/

Can you share exactly what one would have to do to get an app that this applies to and what command exactly to execute to get the error?

File: sigh/lib/assets/resign.sh [hunk 1]
```diff
@@ -69,6 +69,9 @@
 # new features March 2019
 # 1. two more fixes for only creating the archived-expanded-entitlements.xcent file if the version of Xcode < 9.3 as Xcode 10 does not create it.
 #
+# new features August 2019
+# 1. skip creating app entitlements if --use-app-entitlements is used and the app has none. May happen for widgets or other nested apps.
```
Comment (joshdholtz) [inline sigh/lib/assets/resign.sh]: Thanks for adding this 😊

**Review by joshdholtz — APPROVED** [review]
This looks good to me! Might be good to also get some 👀 from either @mgrebenets or @lyndsey-ferguson if they don't mind 😇 I believe they have also had recent PRs or reviews on resign

**Review by lyndsey-ferguson — CHANGES_REQUESTED** [review]
From what I understand of this change, it is simply using the block of code to replace the entitlements and bundle ids when use application entitlements is requested and there are no actual application entitlements. It is not using the app entitlements (as requested) but rather the entitlements stored in the given profile. 😕 

Personally, I feel that this will make it more confusing to future developers who come here.

This script has gotten, in my opinion, unwieldy and difficult. We could accept this change, which would make it more awkward. We could ask @MiB3 to refactor this code significantly, but I don't feel that It is fair to someone who comes in here to fix a legitimate issue to ask that they refactor so.

There must be a middle ground.

@MiB3 would you be open to refactor the bit of code lines 821 to 840 into a new function and call that function there?  Off the cuff, I could imagine that it would take as arguments: an old bundle id, a new bundle id, and a path to an entitlements file.

If so, I would be open to create a few Issues with guidance on how to refactor bits of this script to use the new method.

**Comment by joshdholtz** [comment]
@lyndsey-ferguson Thank you so much for the review! ❤️ 

>This script has gotten, in my opinion, unwieldy and difficult.

☝️ I also agree with this but I feel like it started unwieldy and difficult 😛  I was talking with @janpio and at some point it might be good to maybe even break this out into its own plugin and maybe even rewrite some of this to be more manageable 🤔

**Comment by lyndsey-ferguson** [comment]
@joshdholtz well, as a good start, breaking it up into pieces as we go may be a good start. I'd hate to have that as a TODO item that never gets done...

Once it gets broken into smaller pieces, it is much easier to move it over to its own plugin. Or, moving the script to its own plugin _as is_ could also work. I wouldn't ask @MiB3 to do that though :).

I'm looking for something that can be done in a day or two (depending on the M's workload).

**Comment by mgrebenets** [comment]
@MiB3 Is this something that's broken in Xcode 11? Seems to be new.

Definitely 👍 for moving this whole functionality to a plugin if that's the direction Fastlane is moving towards.

In very brief terms, there'd be 3 major steps
- Move out to a plugin _as is_
- Cover with tests (need some dummy signed apps and such)
- Rewrite in Ruby to make it more clear, testable and manageable

Wouldn't be an easy task though.

**Review by mgrebenets — APPROVED** [review]

**Comment by MiB3** [comment]
Hi everyone, sorry for the silence. I am quite busy at the moment and can therefore not do the "moving to a plugin" (although I really would enjoy doing this). The refactoring from line 821 to 840 should however be doable in the next few weeks.

@janpio I will add the exact commands here in the next days. I tried to figure out how to get the app in a state where the binary of the today extension is signed but does not include any entitlements. However I did not manage to get this from scratch. I only now that with one of my apps I do have this issue. If I find out how to create such an app I will post it here.

@mgrebenets No this is not Xcode 11 related. I used Xcode 10.1 to create the app I now want to resign with new (valid) provisioining profiles.

**Comment by MiB3** [comment]
@janpio 
The command I execute is the following:

```
fastlane sigh resign 
	app_to_resign.ipa 
	--signing_identity $Name_of_signing_identity 
	--provisioning_profile ad_hoc_provisioing_profile_for_main_app.mobileprovision 
	--keychain_path keychain_containing_signing_identity.keychain
	--provisioning_profile $bundle_identifier_of_today_extension=ad_hoc_provisioing_profile_for_today_extension.mobileprovision
	--use_app_entitlements 
```
which internally calls:
```
fastlane/sigh/lib/assets/resign.sh 
	app_to_resign.ipa 
	$Cert_identity 
	-p ad_hoc_provisioing_profile_for_main_app.mobileprovision 
	-p $bundle_identifier_of_today_extension=ad_hoc_provisioing_profile_for_today_extension.mobileprovision      
	--use-app-entitlements   
	--keychain-path keychain_containing_signing_identity.keychain 
	app_to_resign.ipa
```
then the following error ouput appears:

```
/usr/bin/codesign -d --entitlements :'_floatsignTemp/appEntitlements' '_floatsignTemp/Payload/app_to_resign.app/PlugIns/today_extension.appex'
Executable=/Users/bombschm/Desktop/resigning/_floatsignTemp/Payload/app_to_resign.app/PlugIns/today_extension.appex/today_extension
fastlane/sigh/lib/assets/resign.sh: line 618: abort: command not found
Cannot parse a NULL or zero-length data
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
_floatsignTemp/patchedEntitlements: Could not modify plist, error: Failed to parse value Error Reading File: _floatsignTemp/appEntitlements with type -xml
$Cert_identity: no identity found
Encountered an error, aborting!
```

The problem is that the binary `app_to_resign.app/PlugIns/today_extension.appex/today_extension` 
does not include any entitlements.

`/usr/bin/codesign -d --entitlements :'_floatsignTemp/appEntitlements' '_floatsignTemp/Payload/app_to_resign.app/PlugIns/today_extension.appex'`
therefore creates an empty 'appEntitlements' file.

**Comment by MiB3** [comment]
@lyndsey-ferguson I refactored the lines around 830 which made the most sense. I don't think it helps to just put 821 to 840 into one function. The way I did it we can at least reuse the function 2 more times. 
However I am not convinced that this improves the readability of the whole file. And I really do not see an easy way to do so. A rewrite in ruby would probably be best. So we get useful function declarations with named parameters and stuff.

In case you want to merge this, please let me do one last check on my side. I will hopefully do it tomorrow or at least in the next few days.

**Comment by lyndsey-ferguson** [comment]
@MiB3 I'm ecstatic with the change that you've made. For me, this improves readability of the whole file marginally. To improve it fully, the rewrite is necessary, but I would never ask that of someone making a small fix in order for this to merge.

**Comment by lyndsey-ferguson** [comment]
Oh, and I don't have merge authority, but @janpio have been asking my opinion on this as my team and I have had to work a lot with resigning and we have a clue 😆

**Comment by janpio** [comment]
(A rebase on `master` should get rid of the remaining test failures.)

Let us know when you are confident with your changes @MiB3 and I will try to find someone to properly work through them to get this PR merged somehow.

**Comment by MiB3** [comment]
Okey, my test passed! :-)
I can do the rebase on the current master head to get the tests pass again. This will however take a few days again ;-)

**Review by mgrebenets — APPROVED** [review]

**Comment by MiB3** [comment]
Hi everyone, 
I rebased again on master to resolve merge conflicts which occured in the meantime. From my side again everything is fine. Could you please merge this @janpio ? Or do you need something else from me?
@lyndsey-ferguson Could you please approve the Pull-Request? Maybe this is why the maintainers are hestitating to merge ;-)

**Review by lyndsey-ferguson — APPROVED** [review]

**Comment by lyndsey-ferguson** [comment]
I didn't know I had so much power. I feel dangerous now 😄

**Comment by MiB3** [comment]
Hey guys, any chance this will get merged soon?

**Comment by csc-EricWu** [comment]
Is there any update here?

**Comment by MiB3** [comment]
Hi @csc-EricWu I did not realize there were merge conflicts in the mean time. I will try to fix those. Let's say in the next 2 weeks.

**Comment by MiB3** [comment]
Okey, I close this PR, because I don't have the time to fix the merge conflicts and go through the testing on my side again.
The Problem originally probably occured, because of a not optimal build process on my side. This changed in the meantime and the Problem does not occure with the newest version of my app.
If this comes up again I will open a new PR and hopefully can provide steps to reproduce the problem 🙂

**Pull request closed without merging**